### PR TITLE
Implement s3 support for MLFLow plugin

### PR DESCRIPTION
**Pull request by albertoperdomo2**
Implement s3 support for the MLFlow deployment. Since many of the users of NVIDIA Triton Inference Server use the (supported) AWS s3 cloud object storage, this PR builds s3 support for the MLFlow plugin for those times in which the server uses an s3 bucket as a model registry. 

The introduced changes are backwards compatible.

### CODE REVIEW

**Comment by albertoperdomo2** [comment]
@dyastremsky by any chance could you assign a reviewer to get some feedback? TIA!

File: deploy/mlflow-triton-plugin/mlflow_triton/config.py
Comment (dyastremsky) [inline deploy/mlflow-triton-plugin/mlflow_triton/config.py:39]: Could you make sure there is a newline at the end of the file?

**Comment by dyastremsky** [comment]
Hi Alberto! Thank you for contributing this PR and tagging me. I can assign a reviewer.

Can you please sign and submit your [Contributor License Agreement](https://github.com/triton-inference-server/server/blob/main/CONTRIBUTING.md#contributor-license-agreement-cla) first, if you haven't yet?

File: deploy/mlflow-triton-plugin/README.md
Comment (dyastremsky) [inline deploy/mlflow-triton-plugin/README.md:77]: Please shorten this line to under 100 characters (break into multiple lines).

**Comment by albertoperdomo2** [comment]
@dyastremsky just sent the CLA! 👍🏽

**Comment by dyastremsky** [comment]
Thanks, Alberto! This will get reviewed soon.

**Comment by jackroseman** [comment]
Can we get a reviewer assigned? This feature is much needed. Thanks!

**Review by jackroseman — APPROVED** [review]

**Review by jackroseman — DISMISSED** [review]

**Review by GuanLuo — APPROVED** [review]
LGTM. Do you have corresponding test case that can be added to qa/L0_mlflow?

One suggestion that is not require and may be done in a separate PR: You can abstract the file system operations into classes, i.e. S3FileSystem and LocalFileSystem that inherit from FileSystem abstract class, then the plugin only needs to create the proper object during initialization, and the rest of the API can interact with the FileSystem APIs.

**Comment by albertoperdomo2** [comment]
@GuanLuo I'll look into the test and add it to this PR. 

Regarding the other suggestion, sounds reasonable to me, but as you said, that can be added in a separate PR and it will be nice to have this changes merged ASAP, so I will prepare another PR with those changes as soon as this one is merged. 

Btw, thanks for the review!

**Comment by GuanLuo** [comment]
Let me know when the test is added and ready for another round of review

**Comment by albertoperdomo2** [comment]
@GuanLuo I'm facing a few issue while building the QA image, but I hope to have it sorted out today or tomorrow.

**Comment by albertoperdomo2** [comment]
Waiting for #5585. Any help to move this forward will be highly appreciated!

**Comment by albertoperdomo2** [comment]
I have added what I considered as appropriate tests. @GuanLuo @jackroseman

**Comment by GuanLuo** [comment]
Thanks, let me triggered an CI to verify the change. Will keep you posted

File: qa/L0_mlflow/plugin_test.py
Comment (GuanLuo) [inline qa/L0_mlflow/plugin_test.py:141]: I am looking into making this an valid test. The issue is that `TRITON_MODEL_REPO` should also be the path that Triton poll the model from, which should be a valid S3 path. The mock here only overwrite the client side of things and the server will try to poll from the path expecting it to be a valid S3 path. I think the `test.sh` need to be modified to properly setup a S3 bucket and the `plugin_test.py` shouldn't need to change, running the existing tests (`test_onnx_xxx`) should be sufficient as the plugin should internally detect `TRITON_MODEL_REPO` to be a S3 location and upload the files accordingly

Comment (albertoperdomo2) [reply qa/L0_mlflow/plugin_test.py:141]: I still think we need to mock the s3 connection, otherwise the boto3 client will try to connect to an s3 bucket that is not mocked nor exists. Maybe we just need a pre-action for the s3 test that uploads the model to the mocked s3 environment so the model can be pulled from TRITON_MODEL_REPO. What do you think?

Comment (GuanLuo) [reply qa/L0_mlflow/plugin_test.py:141]: Note that Triton server is running along side with the plugin, so Triton server will try to poll `TRITON_MODEL_REPO` which is not an actual bucket on S3 (it is actually local), so Triton can't properly poll the model files and load the model, unless the same mock is injected in Triton server.

Also, the added test is not relaunching the Triton server to monitor the S3 model repository, so it appears to pass because the previous tests have coincidentally setup local model repository with the same model name `onnx_model` and passed the validation.

I have updated the test to set up Triton for S3 separately in this commit https://github.com/triton-inference-server/server/commit/86fb33ca050173680c7f73c8fca0d9daa253e8bd, do you mind to adopt that and test the plugin changes? I ran it locally and it seems like the test failed because the plugin only uploaded part of the model files (only `mlflow-meta.json` and the actual `model.onnx` is missing):
```
# server log
E0411 19:01:34.283724 11677 model_repository_manager.cc:1245] Poll failed for model directory 'onnx_model': Invalid model name: Could not determine backend for model 'onnx_model' with no backend in model configuration. Expected model name of the form 'model.<backend_name>'.

# bucket content
aws s3 rm s3://triton-bucket-gluo-mlflow/ --recursive --include '*'
delete: s3://triton-bucket-gluo-mlflow/onnx_model_with_files/mlflow-meta.json
delete: s3://triton-bucket-gluo-mlflow/onnx_model/mlflow-meta.json
```

Comment (GuanLuo) [reply qa/L0_mlflow/plugin_test.py:141]: To run the test local, you need to set the s3 environment variables `AWS_DEFAULT_REGION`, `AWS_ACCESS_KEY_ID` `AWS_SECRET_ACCESS_KEY`, and provide `CI_JOB_ID` to construct the bucket name used in the test before running the bash script.

Comment (albertoperdomo2) [reply qa/L0_mlflow/plugin_test.py:141]: > Note that Triton server is running along side with the plugin, so Triton server will try to poll TRITON_MODEL_REPO which is not an actual bucket on S3 (it is actually local), so Triton can't properly poll the model files and load the model, unless the same mock is injected in Triton server.

That's right, I didn't think about it. I'll include the commit you shared (btw, thanks!) and will try to see why it fails. I haven't been able to run the CI yet, still facing several issues, but maybe I can try to run the test bash script if I can pull the needed resources to run the test. I'll let you know ASAP!

Comment (albertoperdomo2) [reply qa/L0_mlflow/plugin_test.py:141]: I have managed to set up a local QA environment. I think I am getting more or less where the issue comes from: I tested this with a Tensorflow model and `triton` flavor, which assumes the model is already in the format/structure that Triton expects, that is: 
```
model
├── 1
│   └── model.savedmodel
│       ├── assets
│       ├── keras_metadata.pb
│       ├── saved_model.pb
│       └── variables
│           ├── variables.data-00000-of-00001
│           └── variables.index
└── config.pbtxt
```

The issue with onnx is that it works in a different way which I didn't test. I'm working on it now so it can get solved ASAP! 

Thanks so much for your additions to `test.sh`, it made it really easy to test this changes! I'll keep you posted.

**Comment by albertoperdomo2** [comment]
@GuanLuo ready for review! Test passed on my side.

**Review by GuanLuo — APPROVED** [review]
@albertoperdomo2 Thanks, the change looks good, kicked off another CI. Will merge once passed if there are no further changes from your side.

**Comment by GuanLuo** [comment]
Merged, thanks for your contribution!

**Comment by albertoperdomo2** [comment]
@jackroseman the plugin reads from your MLFlow model registry, downloads the model locally, uploads it to s3 and deletes the local copy. Can you share the mlflow deployments command line that your are running? Thanks!

**Comment by jackroseman** [comment]
@albertoperdomo2 I am using the python API: 
```
client = get_deploy_client('triton')

def deploy_mlflow_model(model_name, version):
    model_uri = f"models:/{model_name}/{version}"
    try:
        print(f"Attempting to create deployment {model_name} v{version}")
        client.create_deployment(model_name, model_uri, flavor="onnx")
    except Exception as e:
        print(f"Error deploying {model_name} v{version}: {e}")

deploy_mlflow_model(
    model_name='my_model',
    version='1'
)
```

When run the above code, the following logs:
```
Attempting to create deployment my_model v1
Copied /tmp/tmpc5fl646v/model.onnx to https://<hidden hostname>:<hidden port num>/triton-mlflow-plugin-model-repo/my_model/1
Saved mlflow-meta.json to https://<hidden hostname>:<hidden port num>/triton-mlflow-plugin-model-repo/my_model
Error deploying my_model v1: failed to load 'my_model', failed to poll from model repository
```
Then on the server running in explicit mode, I see:
```
I0420 16:49:47.866128 1 http_server.cc:3216] HTTP request: 2 /v2/repository/index
I0420 16:49:47.866203 1 filesystem.cc:2335] Using credential    for path  s3://<hidden hostname>:<hidden port num>/triton-mlflow-plugin-model-repo
I0420 16:49:48.678466 1 model_lifecycle.cc:264] ModelStates()
I0420 16:49:50.535424 1 http_server.cc:3216] HTTP request: 2 /v2/repository/models/my_model/load
I0420 16:49:50.535494 1 filesystem.cc:2335] Using credential    for path  s3://<hidden hostname>:<hidden port num>/triton-mlflow-plugin-model-repo/my_model
```
I would like to point out that at the beginning the model repository is empty and when I start the server up it does not try to load any models. 

Let me know what else you need.

**Comment by jackroseman** [comment]
After looking more into this I realized that 's3' was not being set in `client.server_config` and as a result, was not being uploaded to the bucket, But I do not understand why because my TRITON_MODEL_REPO environment variable starts with "s3://".

**Comment by albertoperdomo2** [comment]
Hmm that is really weird, because the Config() is called when the TritonPlugin initialises. Would you be able to add a "debugging" print below [this line](https://github.com/triton-inference-server/server/blob/main/deploy/mlflow-triton-plugin/mlflow_triton/config.py#L39) and re-install the plugin? Maybe something like `print('Triton model repo:' , self['triton_model_repo'])` to see exactly what it prints. The only thing that comes to my mind is that the repo/bucket has some precedent space or something, but it shouldn't be the case.

**Pull request merged**
